GitException: An object of type commit could not be located at offset 60529

Nerdbank GitVersioning has been working for a couple of years with Gitlab (On-site hosted).
We recently upgraded our .net projects from aspnet core 3.2 to .net 6. and Gitlab has not changed, but we are now seeing issues compiling.

We now see an exclamation icon in Visual Studio 2022 under the dependencies section for NerdBank.GitVersioning (3.5.109)
We also get the following error when trying to compile:

```
Error MSB4018: The "Nerdbank.GitVersioning.Tasks.GetBuildVersion" task failed unexpectedly.
Nerdbank.GitVersioning.GitException: An object of type commit could not be located at offset 60529.
   at Nerdbank.GitVersioning.ManagedGit.GitPackReader.GetObject(GitPack pack, Stream stream, Int64 offset, String objectType, GitPackObjectType packObjectType)
   at Nerdbank.GitVersioning.ManagedGit.GitPack.GetObject(Int64 offset, String objectType)
   at Nerdbank.GitVersioning.ManagedGit.GitPack.TryGetObject(GitObjectId objectId, String objectType, Stream& value)
   at Nerdbank.GitVersioning.ManagedGit.GitRepository.TryGetObjectBySha(GitObjectId sha, String objectType, Stream& value)
   at Nerdbank.GitVersioning.ManagedGit.GitRepository.GetObjectBySha(GitObjectId sha, String objectType)
   at Nerdbank.GitVersioning.ManagedGit.GitRepository.GetCommit(GitObjectId sha, Boolean readAuthor)
   at Nerdbank.GitVersioning.Managed.GitExtensions.<GetCommitHeight>g__TryCalculateHeight|4_0(GitCommit commit, <>c__DisplayClass4_0& )
   at Nerdbank.GitVersioning.Managed.GitExtensions.GetCommitHeight(GitRepository repository, GitCommit startingCommit, GitWalkTracker tracker, Func`2 continueStepping)
   at Nerdbank.GitVersioning.Managed.GitExtensions.GetHeight(ManagedGitContext context, Func`2 continueStepping)
   at Nerdbank.GitVersioning.Managed.GitExtensions.GetVersionHeight(ManagedGitContext context, Version baseVersion)
   at Nerdbank.GitVersioning.Managed.ManagedGitContext.CalculateVersionHeight(VersionOptions committedVersion, VersionOptions workingVersion)
   at Nerdbank.GitVersioning.VersionOracle..ctor(GitContext context, ICloudBuild cloudBuild, Nullable`1 overrideVersionHeightOffset)
   at Nerdbank.GitVersioning.Tasks.GetBuildVersion.ExecuteInner()
   at MSBuildExtensionTask.ContextAwareTask.Execute()
   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext() (18, 6)
```

Hi, we've encountered a somewhat similar issue. Our error is:
```
2022-08-24 07:51:57.610 [ERROR] [MsBuild] PrivateP2PCaching.proj: C:\Users\builder-prod-pool-2\AppData\Roaming\Nuget\packages\nerdbank.gitversioning\3.5.108\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018: The "Nerdbank.GitVersioning.Tasks.GetBuildVersion" task failed unexpectedly.
2022-08-24 07:51:57.610 [ERROR] [MsBuild] C:\Users\...\AppData\Roaming\Nuget\packages\nerdbank.gitversioning\3.5.108\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018: System.InvalidOperationException: Unable to get version from commit: e7e021712bf243258ed66fc0a0bd505ea6f5ab49
2022-08-24 07:51:57.610 [ERROR] [MsBuild] C:\Users\...\AppData\Roaming\Nuget\packages\nerdbank.gitversioning\3.5.108\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:  ---> System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
2022-08-24 07:51:57.610 [ERROR] [MsBuild] C:\Users\...\AppData\Roaming\Nuget\packages\nerdbank.gitversioning\3.5.108\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.ManagedGit.GitTreeStreamingReader.FindNode(Stream stream, ReadOnlySpan`1 name)
2022-08-24 07:51:57.610 [ERROR] [MsBuild] C:\Users\...\AppData\Roaming\Nuget\packages\nerdbank.gitversioning\3.5.108\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.ManagedGit.GitRepository.GetTreeEntry(GitObjectId treeId, ReadOnlySpan`1 nodeName)
2022-08-24 07:51:57.610 [ERROR] [MsBuild] C:\Users\...\AppData\Roaming\Nuget\packages\nerdbank.gitversioning\3.5.108\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.Managed.ManagedVersionFile.GetVersion(GitCommit commit, String repoRelativeProjectDirectory, Dictionary`2 blobVersionCache, String& actualDirectory)
2022-08-24 07:51:57.610 [ERROR] [MsBuild] C:\Users\...\AppData\Roaming\Nuget\packages\nerdbank.gitversioning\3.5.108\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.Managed.GitExtensions.GitWalkTracker.GetVersion(GitCommit commit)
2022-08-24 07:51:57.610 [ERROR] [MsBuild] C:\Users\...\AppData\Roaming\Nuget\packages\nerdbank.gitversioning\3.5.108\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    --- End of inner exception stack trace ---
2022-08-24 07:51:57.610 [ERROR] [MsBuild] C:\Users\...\AppData\Roaming\Nuget\packages\nerdbank.gitversioning\3.5.108\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.Managed.GitExtensions.GitWalkTracker.GetVersion(GitCommit commit)
2022-08-24 07:51:57.610 [ERROR] [MsBuild] C:\Users\...\AppData\Roaming\Nuget\packages\nerdbank.gitversioning\3.5.108\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.Managed.GitExtensions.CommitMatchesVersion(GitCommit commit, SemanticVersion expectedVersion, Position comparisonPrecision, GitWalkTracker tracker)
2022-08-24 07:51:57.610 [ERROR] [MsBuild] C:\Users\...\AppData\Roaming\Nuget\packages\nerdbank.gitversioning\3.5.108\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.Managed.GitExtensions.<>c__DisplayClass1_0.<GetVersionHeight>b__0(GitCommit c)
2022-08-24 07:51:57.610 [ERROR] [MsBuild] C:\Users\...\AppData\Roaming\Nuget\packages\nerdbank.gitversioning\3.5.108\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.Managed.GitExtensions.<GetCommitHeight>g__TryCalculateHeight|4_0(GitCommit commit, <>c__DisplayClass4_0& )
2022-08-24 07:51:57.610 [ERROR] [MsBuild] C:\Users\...\AppData\Roaming\Nuget\packages\nerdbank.gitversioning\3.5.108\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.Managed.GitExtensions.GetCommitHeight(GitRepository repository, GitCommit startingCommit, GitWalkTracker tracker, Func`2 continueStepping)
2022-08-24 07:51:57.610 [ERROR] [MsBuild] C:\Users\...\AppData\Roaming\Nuget\packages\nerdbank.gitversioning\3.5.108\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.Managed.GitExtensions.GetHeight(ManagedGitContext context, Func`2 continueStepping)
2022-08-24 07:51:57.610 [ERROR] [MsBuild] C:\Users\...\AppData\Roaming\Nuget\packages\nerdbank.gitversioning\3.5.108\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.Managed.GitExtensions.GetVersionHeight(ManagedGitContext context, Version baseVersion)
2022-08-24 07:51:57.610 [ERROR] [MsBuild] C:\Users\...\AppData\Roaming\Nuget\packages\nerdbank.gitversioning\3.5.108\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.Managed.ManagedGitContext.CalculateVersionHeight(VersionOptions committedVersion, VersionOptions workingVersion)
2022-08-24 07:51:57.610 [ERROR] [MsBuild] C:\Users\...\AppData\Roaming\Nuget\packages\nerdbank.gitversioning\3.5.108\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.VersionOracle..ctor(GitContext context, ICloudBuild cloudBuild, Nullable`1 overrideVersionHeightOffset)
2022-08-24 07:51:57.610 [ERROR] [MsBuild] C:\Users\...\AppData\Roaming\Nuget\packages\nerdbank.gitversioning\3.5.108\build\Nerdbank.GitVersioning.Inner.targets(17,5): error MSB4018:    at Nerdbank.GitVersioning.Tasks.GetBuildVersion.ExecuteInner()
2022-08-24 07:51:57.610 [ERROR] [MsBuild] >    <Nerdbank.GitVersioning.Tasks.GetBuildVersion
2022-08-24 07:51:57.610 [ERROR] [MsBuild] >    ^
2022-08-24 07:51:57.814 [ERROR] [MsBuild] PrivateP2PCaching.proj (no warnings, 1 error) FAILED TO BUILD in 0.3s
2022-08-24 07:51:57.814 [ERROR] [MsBuild] ... FAILED TO BUILD in 0.8s
```
Unfortunately, our repository is not OSS and we cannot share it. What we've been able to confirm so far:
- libgit2 engine works but managed git implementation doesn't
- running `git gc` "fixes" the problem

@AArnott I can try to attach a debugger and analyze the problem, but I don't know what to look for. Can you give some hints, what to check when the error occurs?

--- a/src/NerdBank.GitVersioning/ManagedGit/GitPackDeltafiedStream.cs
+++ b/src/NerdBank.GitVersioning/ManagedGit/GitPackDeltafiedStream.cs
@@ -135,11 +135,7 @@ namespace Nerdbank.GitVersioning.ManagedGit
             if (origin == SeekOrigin.Begin && offset > this.position)
             {
                 // We can optimise this by skipping over instructions rather than executing them
-                int length = (int)(offset - this.position);
-
-                byte[] buffer = ArrayPool<byte>.Shared.Rent(length);
-                this.Read(buffer, 0, length);
-                ArrayPool<byte>.Shared.Return(buffer);
+                this.ReadExactly(checked((int)(offset - this.position)));
                 return this.position;
             }
             else

--- a/src/NerdBank.GitVersioning/ManagedGit/GitPackMemoryCacheStream.cs
+++ b/src/NerdBank.GitVersioning/ManagedGit/GitPackMemoryCacheStream.cs
@@ -73,13 +73,9 @@ namespace Nerdbank.GitVersioning.ManagedGit
 
             if (offset > this.cacheStream.Length)
             {
-                var toRead = (int)(offset - this.cacheStream.Length);
-                byte[] buffer = ArrayPool<byte>.Shared.Rent(toRead);
-                int read = this.stream.Read(buffer, 0, toRead);
                 this.cacheStream.Seek(0, SeekOrigin.End);
-                this.cacheStream.Write(buffer, 0, read);
-                ArrayPool<byte>.Shared.Return(buffer);
-
+                int toRead = (int)(offset - this.cacheStream.Length);
+                this.stream.ReadExactly(toRead, this.cacheStream);
                 this.DisposeStreamIfRead();
                 return this.cacheStream.Position;
             }

--- a/src/NerdBank.GitVersioning/ManagedGit/StreamExtensions.cs
+++ b/src/NerdBank.GitVersioning/ManagedGit/StreamExtensions.cs
@@ -150,5 +150,32 @@ namespace Nerdbank.GitVersioning.ManagedGit
             return true;
         }
 #endif
+
+        /// <summary>
+        /// Reads the specified number of bytes from a stream, or until the end of the stream.
+        /// </summary>
+        /// <param name="readFrom">The stream to read from.</param>
+        /// <param name="length">The number of bytes to be read.</param>
+        /// <param name="copyTo">The stream to copy the read bytes to, if required.</param>
+        /// <returns>The number of bytes actually read. This will be less than <paramref name="length"/> only if the end of <paramref name="readFrom"/> is reached.</returns>
+        internal static int ReadExactly(this Stream readFrom, int length, Stream? copyTo = null)
+        {
+            int bytesRemaining = length;
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(Math.Min(50 * 1024, bytesRemaining));
+            while (bytesRemaining > 0)
+            {
+                int read = readFrom.Read(buffer, 0, Math.Min(buffer.Length, bytesRemaining));
+                if (read == 0)
+                {
+                    break;
+                }
+
+                copyTo?.Write(buffer, 0, read);
+                bytesRemaining -= read;
+            }
+
+            ArrayPool<byte>.Shared.Return(buffer);
+            return length - bytesRemaining;
+        }
     }
 }

--- a/src/NerdBank.GitVersioning/ManagedGit/ZLibStream.cs
+++ b/src/NerdBank.GitVersioning/ManagedGit/ZLibStream.cs
@@ -151,11 +151,7 @@ namespace Nerdbank.GitVersioning.ManagedGit
             if (origin == SeekOrigin.Begin && offset > this.position)
             {
                 // We may be able to optimize this by skipping over the compressed data
-                int length = (int)(offset - this.position);
-
-                byte[] buffer = ArrayPool<byte>.Shared.Rent(length);
-                this.Read(buffer, 0, length);
-                ArrayPool<byte>.Shared.Return(buffer);
+                this.ReadExactly(checked((int)(offset - this.position)));
                 return this.position;
             }
             else

